source/v1/I2CDetail.cpp

```diff
@@ -0,0 +1,258 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mbed-drivers/v1/I2CDetail.hpp"
+
+#if DEVICE_I2C && DEVICE_I2C_ASYNCH
+
+#include "mbed-drivers/v1/I2C.hpp"
+#include "core-util/CriticalSectionLock.h"
+#include "minar/minar.h"
+
+namespace mbed_drivers {
+using namespace mbed;
+namespace v1 {
+namespace detail {
+
+int I2CResourceManager::post_transaction(I2CTransaction *t) {
+    int rc;
+    if ((rc = validate_transaction(t))) {
+        return rc;
+    }
+    // Optimization NOTE: this could be lock-free with atomic_cas.
+    mbed::util::CriticalSectionLock lock;
+    // Find the tail of the queue
+    I2CTransaction * tx = (I2CTransaction *)TransactionQueue;
+    if (!tx) {
+        TransactionQueue = t;
+        power_up();
+        start_transaction();
+        return 0;
+    }
+    while (tx->next) {
+        tx = tx->next;
+    }
+    tx->next = t;
+    return 0;
+}
+
+void I2CResourceManager::process_event(int event) {
+    I2CTransaction * t;
+    CORE_UTIL_ASSERT(TransactionQueue != nullptr);
+    if (!TransactionQueue) {
+        return;
+    }
+    // Get the current item in the transaction queue
+    t = (I2CTransaction *) TransactionQueue;
+    if (!t) {
+        return;
+    }
+    // Fire the irqcallback for the segment
+    t->current->call_irq_cb();
+    // If there was an event that is not a complete event
+    // or there was a complete event and the next segment is nullptr
+    if ((event & I2C_EVENT_ALL & ~I2C_EVENT_TRANSFER_COMPLETE) ||
+        ((event & I2C_EVENT_TRANSFER_COMPLETE) && t->current->get_next() == nullptr))
+    {
+        // fire the handler
+        using cbtype = mbed::util::FunctionPointer2<void,I2CTransaction *, int>;
+        minar::Scheduler::postCallback(cbtype(this, &I2CResourceManager::event_cb).bind(t,event));
+        // TODO: This could be lock-free
+        // Enter a critical section
+        mbed::util::CriticalSectionLock lock;
+        // Advance to the next transaction
+        TransactionQueue = t->next;
+        if (TransactionQueue) {
+            // Initiate the next transaction
+            start_transaction();
+        } else {
+            power_down();
+        }
+    } else {
+        // TODO: this could be lock-free
```

Comment:
Another instance of a TODO that does;t belong in production code, likewise above.